kafka-client/src/test/java/dev/responsive/kafka/internal/db/rs3/client/grpc/GrpcRS3ClientEndToEndTest.java

```diff
@@ -0,0 +1,249 @@
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+
+import com.google.protobuf.ByteString;
+import dev.responsive.kafka.internal.db.rs3.client.LssId;
+import dev.responsive.kafka.internal.db.rs3.client.Put;
+import dev.responsive.rs3.RS3Grpc;
+import dev.responsive.rs3.Rs3;
+import io.grpc.ClientCall;
+import io.grpc.ManagedChannel;
+import io.grpc.Server;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import io.grpc.inprocess.InProcessChannelBuilder;
+import io.grpc.inprocess.InProcessServerBuilder;
+import io.grpc.stub.StreamObserver;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.kafka.common.utils.Time;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+class GrpcRS3ClientEndToEndTest {
+
+  private static final String SERVER_NAME = "localhost";
+  private static final long RETRY_TIMEOUT_MS = 10000;
+  private static final UUID STORE_ID = UUID.randomUUID();
+  private static final int PSS_ID = 0;
+  private static final LssId LSS_ID = new LssId(PSS_ID);
+
+  private Server server;
+  private ManagedChannel channel;
+  private GrpcRS3Client client;
+
+  @BeforeEach
+  public void setUp() throws IOException {
+    this.server = InProcessServerBuilder
+        .forName(SERVER_NAME)
+        .addService(new TestRs3Service())
+        .build()
+        .start();
+    this.channel = Mockito.spy(InProcessChannelBuilder
+        .forName(SERVER_NAME)
+        .directExecutor()
+        .build());
+    this.client = new GrpcRS3Client(
+        new PssStubsProvider(this.channel),
+        Time.SYSTEM,
+        RETRY_TIMEOUT_MS
+    );
+  }
+
+  @AfterEach
+  public void tearDown() {
+    this.channel.shutdownNow();
+    this.server.shutdownNow();
+    this.client.close();
+  }
+
+  @Test
+  public void shouldPutAndGet() {
+    final var sendRecv = client.writeWalSegmentAsync(
+        STORE_ID,
+        LSS_ID,
+        PSS_ID,
+        Optional.empty(),
+        5L
+    );
+
+    final var key = new byte[] { 0 };
+    final var value = new byte[] { 1 };
+    sendRecv.sender().sendNext(new Put(key, value));
+    sendRecv.sender().finish();
+
+    final var flushedOffset = sendRecv
+        .receiver()
+        .toCompletableFuture()
+        .join();
+    assertThat(flushedOffset, is(Optional.of(5L)));
+
+    final var getResult = client.get(
+        STORE_ID,
+        LSS_ID,
+        PSS_ID,
+        Optional.of(5L),
+        key
+    );
+    assertThat(getResult.isPresent(), is(true));
+    final var resultValue = getResult.get();
+    assertThat(resultValue, equalTo(value));
+  }
+
+  @Test
+  public void shouldRetryPutWithNetworkInterruption() {
```

Comment:
These tests get us a little closer to simulating bad network behavior, but still rely on assumptions about how network errors are raised in the grpc stack.